### PR TITLE
Transaction Sponsorship

### PR DESCRIPTION
**Pull request by aliXsed**
- [x] Add update for user limits and pot limits
- [x] Benchmark using the standard machine

### CODE REVIEW

File: pallets/sponsorship/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+					sponsor: who,
+					sponsorship_type,
+					remained_fee_quota: fee_quota,
+					remained_reserve_quota: reserve_quota,
+				},
+			);
+
+			Self::deposit_event(Event::PotCreated(pot));
+			Ok(())
+		}
```
Comment (ETeissonniere) [inline pallets/sponsorship/src/lib.rs:225]: I think this extrinsic should reserve `fee_quota.saturating_add(reserve_quota)` to:
1. Avoid spamming of this extrinsic
2. Ensure the sponsor has enough funds available for the limits they are setting themselves

File: pallets/sponsorship/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+				let pot_details = maybe_pot_details.as_mut().ok_or(Error::<T>::InUse)?;
+				ensure!(pot_details.sponsor == who, Error::<T>::NoPermission);
+				let users = User::<T>::iter_prefix(pot).count();
+				ensure!(users == 0, Error::<T>::InUse);
+				*maybe_pot_details = None;
+				Ok(())
+			})?;
+			Self::deposit_event(Event::PotRemoved(pot));
+			Ok(())
+		}
```
Comment (ETeissonniere) [inline pallets/sponsorship/src/lib.rs:245]: Following the above thought, this would unreserve remaining coins

**Review by ETeissonniere — CHANGES_REQUESTED** [review]
Looking good! I left some minor changes and thoughts

File: pallets/sponsorship/README.md [hunk 3]
```diff
@@ -0,0 +1 @@
+License: MIT-0
```
[the remark this answers is not in the document]
Comment (aliXsed) [reply pallets/sponsorship/README.md]: @ETeissonniere I added a brief description here instead

File: pallets/sponsorship/src/benchmarking.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+//! Benchmarking setup for pallet-sponsorship
+#![cfg(feature = "runtime-benchmarks")]
+use super::*;
+
+#[allow(unused)]
+use crate::Pallet as Sponsorship;
+use frame_benchmarking::v2::*;
+use frame_support::assert_ok;
+use frame_system::RawOrigin;
```
[the remark this answers is not in the document]
Comment (aliXsed) [reply pallets/sponsorship/src/benchmarking.rs:27]: @ETeissonniere `#[allow(unused)]` is still needed when you compile for testing benchmarks:
`cargo test -p pallet-sponsorship --features runtime-benchmarks`

File: support/src/lib.rs [hunk 5]
```diff
@@ -18,6 +18,76 @@
...
+		self.balance
+	}
+	pub fn limit(&self) -> T {
+		self.limit
+	}
+	/// Returns the amount to add to the balance to reach the limit.
+	pub fn available_margin(&self) -> T {
+		self.limit.saturating_sub(self.balance)
+	}
+}
```
[the remark this answers is not in the document]
Comment (aliXsed) [reply support/src/lib.rs]: Done. Thank you for reminding it.

File: runtimes/eden/src/pallets_util.rs [hunk 6]
```diff
@@ -143,6 +144,45 @@ impl pallet_nodle_uniques::Config for Runtime {
...
 }
 
+#[derive(
+	Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, RuntimeDebug, MaxEncodedLen, scale_info::TypeInfo,
+)]
+pub enum SponsorshipType {
+	AnySafe,
+	Uniques,
+	UniquesMint,
+}
```
[the remark this answers is not in the document]
Comment (aliXsed) [reply runtimes/eden/src/pallets_util.rs]: Yes, open to any suggestions. I have [simplified it](https://github.com/NodleCode/chain/pull/767#:~:text=Verified-,6d2a370,-Merge%20state) so we only have two categories for now. But later when we provide more governance decentralisation we can add a category for governance transactions.

File: pallets/sponsorship/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+		PotRemoved(T::PotId),
+		/// Event emitted when user/users are registered indicating the list of them
+		UsersRegistered(T::PotId, Vec<T::AccountId>),
+		/// Event emitted when user/users are removed indicating the list of them
+		UsersRemoved(T::PotId, Vec<T::AccountId>),
+		/// Event emitted when a sponsor_me call has been successful indicating the reserved amount
+		Sponsored { top_up: BalanceOf<T>, refund: BalanceOf<T> },
+		/// Event emitted when the transaction fee is paid showing the payer and the amount
+		TransactionFeePaid(T::AccountId, BalanceOf<T>),
+	}
```
[the remark this answers is not in the document]
Comment (aliXsed) [reply pallets/sponsorship/src/lib.rs]: Good suggestion. Done through https://github.com/NodleCode/chain/pull/767#:~:text=Verified-,0216972,-Merge%20state

File: pallets/sponsorship/src/benchmarking.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ETeissonniere) [reply pallets/sponsorship/src/benchmarking.rs:27]: Oops, got it!

**Review by ETeissonniere — APPROVED** [review]
lgtm. to be merged once the `try-runtime` checks pass

**Pull request merged**